### PR TITLE
[agg_v2] executor exceptions propagation

### PR DESCRIPTION
**Pull request by igor-aptos**
### Description

- make executor resilient to aggregator exceptions. where possible, skip transactions that are failing
- cleanup errors across parallel and sequential execution
- and discard_failed_blocks

### CODE REVIEW

File: aptos-move/aptos-vm/src/block_executor/mod.rs [hunk 1]
```diff
@@ -88,6 +88,10 @@ impl BlockExecutorTransactionOutput for AptosTransactionOutput {
         Self::new(VMOutput::empty_with_status(TransactionStatus::Retry))
     }
 
+    fn discard_output(discard_code: StatusCode) -> Self {
```
Comment (vusirikala) [inline aptos-move/aptos-vm/src/block_executor/mod.rs:91]: What's the difference between abort and discard? When do we discard instead of abort?

File: aptos-move/aptos-vm/src/block_executor/mod.rs [hunk 2]
```diff
@@ -431,11 +435,8 @@ impl BlockAptosVM {
 
                 Ok(BlockOutput::new(output_vec))
             },
-            Err(BlockExecutionError::FallbackToSequential(e)) => {
-                unreachable!(
-                    "[Execution]: Must be handled by sequential fallback: {:?}",
-                    e
-                )
+            Err(BlockExecutionError::FatalBlockExecutorError(PanicError::CodeInvariantError(err_msg))) => {
+                Err(VMStatus::Error { status_code: StatusCode::DELAYED_MATERIALIZATION_CODE_INVARIANT_ERROR, sub_status: None, message: Some(err_msg) })
```
Comment (vusirikala) [inline aptos-move/aptos-vm/src/block_executor/mod.rs:439]: FatalBlockExecutionError happens only because of aggregators?

File: aptos-move/block-executor/src/errors.rs [hunk 3]
```diff
@@ -2,62 +2,47 @@
...
 
 #[derive(Clone, Debug, PartialEq, Eq)]
 /// Logging is bottlenecked in constructors.
-pub enum IntentionalFallbackToSequential {
-    // The same module access path for module was both read & written during speculative executions.
-    // This may trigger a race due to the Move-VM loader cache implementation, and mitigation requires
-    // aborting the parallel execution pipeline and falling back to the sequential execution.
-    // TODO: provide proper multi-versioning for code (like data) for the cache.
-    ModulePathReadWrite,
+pub enum SequentialBlockExecutionError<E> {
```
Comment (vusirikala) [inline aptos-move/block-executor/src/errors.rs:21]: Comment needs to be updated?

File: aptos-move/block-executor/src/errors.rs [hunk 4]
```diff
@@ -2,62 +2,47 @@
...
-
-        IntentionalFallbackToSequential::ResourceGroupSerializationError
-    }
+    ErrorToReturn(BlockExecutionError<E>)
 }
 
 #[derive(Clone, Debug, PartialEq, Eq)]
 pub enum BlockExecutionError<E> {
-    FallbackToSequential(PanicOr<IntentionalFallbackToSequential>),
+    FatalBlockExecutorError(PanicError),
```
Comment (vusirikala) [inline aptos-move/block-executor/src/errors.rs:30]: Comments on when this error occurs would be helpful.

File: aptos-move/block-executor/src/errors.rs [hunk 5]
```diff
@@ -2,62 +2,47 @@
...
+pub struct ModulePathReadWriteError;
+
+// This is not PanicError because we need to match the error variant to provide a specialized
+// fallback logic if a resource group serialization error occurs.
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub struct ResourceGroupSerializationError;
 
 #[derive(Clone, Debug, PartialEq, Eq)]
 /// Logging is bottlenecked in constructors.
-pub enum IntentionalFallbackToSequential {
```
Comment (vusirikala) [inline aptos-move/block-executor/src/errors.rs:13]: Why is this removed?

File: aptos-move/aptos-vm/src/block_executor/mod.rs (hunk 1 above)
Comment (igor-aptos) [reply aptos-move/aptos-vm/src/block_executor/mod.rs:91]: discard means rejected from onchain, i.e. we couldn't even charge the signer gas (for example if sequence number is incorrect, or signature is incorrect, etc)

abort means execution of the transaction failed, but transaction itself gets committed onchain with that status

File: aptos-move/aptos-vm/src/block_executor/mod.rs (hunk 2 above)
Comment (igor-aptos) [reply aptos-move/aptos-vm/src/block_executor/mod.rs:439]: aggregators or resource groups (i.e. delayed materialization)

File: aptos-move/block-executor/src/errors.rs [hunk 6]
```diff
@@ -2,62 +2,47 @@
...
+// This may trigger a race due to the Move-VM loader cache implementation, and mitigation requires
+// aborting the parallel execution pipeline and falling back to the sequential execution.
+// TODO: provide proper multi-versioning for code (like data) for the cache.
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub struct ModulePathReadWriteError;
+
+// This is not PanicError because we need to match the error variant to provide a specialized
+// fallback logic if a resource group serialization error occurs.
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub struct ResourceGroupSerializationError;
```
[the remark this answers is not in the document]
Comment (igor-aptos) [reply aptos-move/block-executor/src/errors.rs:17]: yeah, this is purely for the serialize_groups function, because it is called both in sequential and parallel.

I wanted to make every function return only errors it can create.

File: aptos-move/block-executor/src/task.rs [hunk 7]
```diff
@@ -163,6 +163,9 @@ pub trait TransactionOutput: Send + Sync + Debug {
     /// Execution output for transactions that comes after SkipRest signal.
     fn skip_output() -> Self;
 
+    /// Execution output for transactions that should be discarded.
+    fn discard_output(discard_code: StatusCode) -> Self;
```
Comment (vusirikala) [inline aptos-move/block-executor/src/task.rs:167]: Is discarding different from skipping rest of transactions in the block? A comment on when we discard/skip transactions would be helpful.

File: aptos-move/block-executor/src/proptest_types/baseline.rs [hunk 8]
```diff
@@ -377,7 +377,7 @@ impl<K: Debug + Hash + Clone + Eq> BaselineOutput<K> {
                 assert_eq!(*idx, self.read_values.len());
                 assert_eq!(*idx, self.resolved_deltas.len());
             },
-            Err(BlockExecutionError::FallbackToSequential(_)) => {
+            Err(BlockExecutionError::FatalBlockExecutorError(_)) => {
                 // Parallel execution currently returns an arbitrary error to fallback.
```
Comment (vusirikala) [inline aptos-move/block-executor/src/proptest_types/baseline.rs:381]: Comment may need to be updated.

File: aptos-move/block-executor/src/errors.rs (hunk 5 above)
Comment (igor-aptos) [reply aptos-move/block-executor/src/errors.rs:13]: split it out. parallel only returns FallbackFromParallel, while sequential has others

File: aptos-move/block-executor/src/errors.rs [hunk 9]
```diff
@@ -2,62 +2,47 @@
...
-    pub(crate) fn resource_group_serialization_error(error_msg: String, txn_idx: TxnIndex) -> Self {
-        alert!("[Execution] At txn {}, {:?}", txn_idx, error_msg);
-
-        IntentionalFallbackToSequential::ResourceGroupSerializationError
-    }
+    ErrorToReturn(BlockExecutionError<E>)
 }
 
 #[derive(Clone, Debug, PartialEq, Eq)]
 pub enum BlockExecutionError<E> {
```
[the remark this answers is not in the document]
Comment (igor-aptos) [reply aptos-move/block-executor/src/errors.rs:29]: parallal has only one error type without any metadata - so just used ()

BlockExecutionError is what caller sees. SequentialBlockExecutionError and () is internal , that we process before returning to the caller.

File: aptos-move/block-executor/src/errors.rs [hunk 10]
```diff
@@ -2,62 +2,47 @@
...
-use aptos_mvhashmap::types::TxnIndex;
 use aptos_types::delayed_fields::PanicError;
-use aptos_vm_logging::{alert, prelude::*};
+
+// The same module access path for module was both read & written during speculative executions.
+// This may trigger a race due to the Move-VM loader cache implementation, and mitigation requires
+// aborting the parallel execution pipeline and falling back to the sequential execution.
+// TODO: provide proper multi-versioning for code (like data) for the cache.
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub struct ModulePathReadWriteError;
```
Comment (gelash) [inline aptos-move/block-executor/src/errors.rs:12]: nit: can this (and ResourceGroupSerializationError) be pub(crate)?

File: aptos-move/block-executor/src/errors.rs (hunk 4 above)
Comment (gelash) [reply aptos-move/block-executor/src/errors.rs:30]: Nice, I like the distinction here between fatal block vs VM error

File: aptos-move/block-executor/src/executor.rs [hunk 11]
```diff
@@ -1271,75 +1259,73 @@ where
...
+                // All logs from the first pass of sequential execution should be cleared and not reported.
                 // Clear by re-initializing the speculative logs.
                 init_speculative_logs(signature_verified_block.len());
 
-                ret = self.execute_transactions_sequential(
+                let sequential_result = self.execute_transactions_sequential(
                     executor_arguments,
                     signature_verified_block,
                     base_view,
                     false,
```
Comment (gelash) [inline aptos-move/block-executor/src/executor.rs:1303]: true? and don't we have some failpoint test for this? I guess not (or did it break)?

File: aptos-move/block-executor/src/executor.rs [hunk 12]
```diff
@@ -1271,75 +1259,73 @@ where
...
-                // All logs from the parallel execution should be cleared and not reported.
+        // If sequential gave us result, return it
+        let sequential_error = match sequential_result {
+            Ok(output) => {
+                return Ok(output);
+            },
+            Err(SequentialBlockExecutionError::ResourceGroupSerializationError) => {
+                // All logs from the first pass of sequential execution should be cleared and not reported.
                 // Clear by re-initializing the speculative logs.
                 init_speculative_logs(signature_verified_block.len());
```
Comment (gelash) [inline aptos-move/block-executor/src/executor.rs:1297]: I think if there is logging specifically in the sequential execution logic, it may not be using speculative logging macros (since it never had a fallback). Let's add a TODO(clean-up): to check it later? (I'm happy to audit that at some point, pretty low pri most likely, but still)

File: aptos-move/block-executor/src/executor.rs [hunk 13]
```diff
@@ -1271,75 +1259,73 @@ where
...
+            }
+        };
 
-        // If after trying available fallbacks, we still are askign to do a fallback,
-        // something unrecoverable went wrong.
-        if let Err(BlockExecutionError::FallbackToSequential(e)) = &ret {
-            // TODO[agg_v2][fix] make sure this can never happen - we have sequential raising
-            // this error often when something that should never happen goes wrong
-            panic!("Sequential execution failed with {:?}", e);
+        // if not skip failed blocks enabled
```
Comment (gelash) [inline aptos-move/block-executor/src/executor.rs:1324]: so currently we don't have a config for it and it's not enabled?

File: aptos-move/block-executor/src/executor.rs [hunk 14]
```diff
@@ -1271,75 +1259,73 @@ where
...
+            );
 
-        let mut resource_group_bcs_fallback = matches!(
-            ret,
-            Err(BlockExecutionError::FallbackToSequential(PanicOr::Or(
-                IntentionalFallbackToSequential::ResourceGroupSerializationError
-            )))
-        );
+            // If parallel gave us result, return it
+            if let Ok(output) = parallel_result {
```
Comment (gelash) [inline aptos-move/block-executor/src/executor.rs:1270]: I like early returning in this method, that did simplify a logic a lot, thanks!

File: aptos-move/block-executor/src/executor.rs [hunk 15]
```diff
@@ -1029,6 +1020,24 @@ where
...
+                    if let Some(commit_hook) = &self.transaction_commit_hook {
+                        commit_hook.on_execution_aborted(idx as TxnIndex);
+                    }
+                    alert!("Fatal VM error by transaction {}", idx as TxnIndex);
+                    // Record the status indicating the unrecoverable VM failure.
+                    return Err(SequentialBlockExecutionError::ErrorToReturn(BlockExecutionError::FatalVMError(err)));
+                },
+                ExecutionStatus::DelayedFieldsCodeInvariantError(msg) => {
+                    alert!("Discarding transaction because we got DelayedFieldsCodeInvariantError in the sequential fallback: {msg}");
+                    ret.push(E::Output::discard_output(StatusCode::DELAYED_MATERIALIZATION_CODE_INVARIANT_ERROR));
```
Comment (gelash) [inline aptos-move/block-executor/src/executor.rs:1033]: We are changing the behavior here, and I'm not sure it's what we want to do? Resource group serialization is even more unlikely than these I think and it is also quite certainly constrained to a single transaction. If we see delayed field materialization issue, there is a bigger chance the whole block might have an issue, so returning with a block error seems safer to me (esp if we have a way to skip a block?)

File: aptos-move/block-executor/src/executor.rs [hunk 16]
```diff
@@ -1029,6 +1020,24 @@ where
...
+                    continue;
+                },
+                ExecutionStatus::SpeculativeExecutionAbortError(msg) => {
+                    alert!("Discarding transaction because we got SpeculativeExecutionAbortError in the sequential fallback {msg}");
+                    ret.push(E::Output::discard_output(StatusCode::DELAYED_MATERIALIZATION_CODE_INVARIANT_ERROR));
```
Comment (gelash) [inline aptos-move/block-executor/src/executor.rs:1038]: a different argument here, but also for considering block error instead of discarding would be that this should be so impossible during sequential execution that one could almost advocate unreachable or skipping a block vs discarding an individual transaction.

File: aptos-move/block-executor/src/executor.rs [hunk 17]
```diff
@@ -570,9 +568,6 @@ where
                         txn_idx + 1,
                         scheduler.num_txns(),
                     );
-                    fail_point!("commit-all-halt-err", |_| Err(PanicOr::Or(
```
Comment (gelash) [inline aptos-move/block-executor/src/executor.rs:573]: I think we can remove this test and failpoint as is, what it currently tests (fallback when serialization error happens) is tested by another test unit test. I think it was supposed to pause one thread, so the block finished, and then make sure the error was still propagated. with the current flow, this is easier to see (no halt_ok and halt_err), but I can bring back the proper test later. Maybe add a TODO for me in unit_tests/mod.rs to write a proper test when you remove / disable https://github.com/aptos-labs/aptos-core/blob/ce593c4b0c9f066c821b80c403d6b5f8ec12263b/aptos-move/block-executor/src/unit_tests/mod.rs#L158

File: aptos-move/aptos-vm/src/block_executor/mod.rs (hunk 1 above)
Comment (georgemitenkov) [reply aptos-move/aptos-vm/src/block_executor/mod.rs:91]: It was not the case before that we would skip from BlockSTM. VM can discard, it creates empty outputs already (e.g., if prologue fails)

File: aptos-move/aptos-vm/src/block_executor/mod.rs (hunk 2 above)
Comment (georgemitenkov) [inline aptos-move/aptos-vm/src/block_executor/mod.rs:439]: cannot we reuse `code_invariant_error` here?

Comment (igor-aptos) [reply aptos-move/aptos-vm/src/block_executor/mod.rs:439]: we have From for StatusCode, but not for VMStatus itself?

File: aptos-move/block-executor/src/proptest_types/baseline.rs [hunk 18]
```diff
@@ -230,154 +230,166 @@ impl<K: Debug + Hash + Clone + Eq> BaselineOutput<K> {
         }
     }
 
-    // Used for testing, hence the function asserts the correctness conditions within
-    // itself to be easily traceable in case of an error.
-    pub(crate) fn assert_output<E: Debug>(
-        &self,
-        results: &BlockExecutionResult<BlockOutput<MockOutput<K, E>>, usize>,
-    ) {
+    fn assert_success<E: Debug>(&self, block_output: &BlockOutput<MockOutput<K, E>>) {
```
Comment (igor-aptos) [inline aptos-move/block-executor/src/proptest_types/baseline.rs:233]: git is showing changes here because of indentation, this function is basically the same code as it was before, it's just extracted to be able to be used in both assert_output and assert_parallel_output

File: aptos-move/block-executor/src/executor.rs [hunk 19]
```diff
@@ -653,7 +648,15 @@ where
         let finalized_groups = last_input_output.take_finalized_group(txn_idx);
         let materialized_finalized_groups =
             map_id_to_values_in_group_writes(finalized_groups, &latest_view)?;
-        let serialized_groups = serialize_groups::<T>(materialized_finalized_groups, txn_idx)?;
+
+        fail_point!("commit-all-halt-err", |_| Err(code_invariant_error(
```
Comment (gelash) [inline aptos-move/block-executor/src/executor.rs:652]: I remember why this fail point was where it was!
the idea was to make sure that even if we execute the block successfully, and halt the scheduler, we can still return with error from parallel execution.
Moving it here makes the error be returned from both transactions (the test has 2), so it is possible to return it even before the whole block is prepared for commit. Maybe we can add a check that txn_idx == scheduler.num_txns() - 1?

File: aptos-move/block-executor/src/executor.rs (hunk 17 above)
Comment (gelash) [reply aptos-move/block-executor/src/executor.rs:573]: Recalled why it was here, the test was actually working as desired, as this would only be invoked after the whole block was prepared for commit and scheduler was halted. It's okay to move it to the new place if we add a conditional to fail point return - or we can keep it here - added the comment to the new place.

File: aptos-move/block-executor/src/executor.rs [hunk 20]
```diff
@@ -1271,81 +1278,87 @@ where
...
+                    },
+                    Err(SequentialBlockExecutionError::ErrorToReturn(err)) => err,
+                }
+            },
+            Err(SequentialBlockExecutionError::ErrorToReturn(err)) => err,
+        };
+
+        // if not skip failed blocks enabled
+        if self.config.local.discard_failed_blocks {
+            // We cannot execute block, disard everything (including block metadata and validator transactions)
```
Comment (gelash) [inline aptos-move/block-executor/src/executor.rs:1351]: typo: discard.

Is it safe to discard blockmetadata or some other transactions? (can they have some specialized logic that wouldn't be in the next block?)

File: aptos-move/block-executor/src/executor.rs [hunk 21]
```diff
@@ -1271,81 +1278,87 @@ where
...
+                            "resource group serialization during bcs fallback should not happen",
+                        ))
+                    },
+                    Err(SequentialBlockExecutionError::ErrorToReturn(err)) => err,
+                }
+            },
+            Err(SequentialBlockExecutionError::ErrorToReturn(err)) => err,
+        };
+
+        // if not skip failed blocks enabled
```
Comment (gelash) [inline aptos-move/block-executor/src/executor.rs:1349]: comment seems to not fit anymore?

**Review by gelash — APPROVED** [review]

File: config/src/config/execution_config.rs [hunk 22]
```diff
@@ -67,6 +69,7 @@ impl Default for ExecutionConfig {
             num_proof_reading_threads: 32,
             paranoid_type_verification: true,
             paranoid_hot_potato_verification: true,
+            discard_failed_blocks: false,
```
Comment (georgemitenkov) [inline config/src/config/execution_config.rs:72]: Is it ok to have as false by default? The cell in AptosVM has default of true, isn't this inconsistent?

**Review by georgemitenkov — APPROVED** [review]

File: aptos-move/block-executor/src/executor.rs [hunk 23]
```diff
@@ -833,7 +836,7 @@ where
         executor_initial_arguments: E::Argument,
         signature_verified_block: &[T],
         base_view: &S,
-    ) -> Result<BlockOutput<E::Output>, PanicOr<IntentionalFallbackToSequential>> {
+    ) -> Result<BlockOutput<E::Output>, ()> {
```
[the remark this answers is not in the document]
Comment (igor-aptos) [reply aptos-move/block-executor/src/executor.rs:839]: We have variety of codification that end parallel execution, and we coalesce them all to have it deterministic (because the real causing condition is hard to get).

So it really is a signal "parallel failure" without more context.

File: aptos-move/block-executor/src/executor.rs [hunk 24]
```diff
@@ -1271,81 +1278,87 @@ where
...
+            // If parallel gave us result, return it
+            if let Ok(output) = parallel_result {
+                return Ok(output);
+            }
 
-                if let Err(BlockExecutionError::FallbackToSequential(_)) = &ret {
-                    // Any error logs are already written at appropriate levels.
-                    debug!("Sequential_fallback occurred");
+            if !self.config.local.allow_fallback {
+                panic!("Parallel execution failed and fallback is not allowed");
```
[the remark this answers is not in the document]
Comment (igor-aptos) [reply aptos-move/block-executor/src/executor.rs:1294]: For tests for example, where we are testing correctness of parallel execution,  we must certainly don't want to fallback. 
In prod it is true, we don't have it in config either

File: config/src/config/execution_config.rs (hunk 22 above)
Comment (igor-aptos) [reply config/src/config/execution_config.rs:72]: Ooops, default there to true is wrong, I wanted false. This is SEV mitigation flag

**Review by vusirikala — APPROVED** [review]
